dpu-cni/pkgs/sriovutils/pci_allocator.go

```diff
@@ -2,91 +2,57 @@ package sriovutils
 
 import (
 	"fmt"
-	"os"
-	"path/filepath"
-
 	"github.com/containernetworking/plugins/pkg/ns"
 )
 
-type PCIAllocation interface {
-	SaveAllocatedPCI(string, string) error
-	DeleteAllocatedPCI(string) error
-	IsAllocated(string) error
-}
-
 type PCIAllocator struct {
-	dataDir string
+	store KeyValueStore
 }
 
-// NewPCIAllocator returns a new PCI allocator
-// it will use the <dataDir>/pci folder to store the information about allocated PCI addresses
-func NewPCIAllocator(dataDir string) *PCIAllocator {
-	return &PCIAllocator{dataDir: filepath.Join(dataDir, "pci")}
+func NewPCIAllocator(store KeyValueStore) *PCIAllocator {
+	return &PCIAllocator{
+		store: store,
+	}
 }
 
-// SaveAllocatedPCI creates a file with the pci address as a name and the network namespace as the content
-// return error if the file was not created
 func (p *PCIAllocator) SaveAllocatedPCI(pciAddress, ns string) error {
-	if err := os.MkdirAll(p.dataDir, 0600); err != nil {
-		return fmt.Errorf("failed to create the sriov data directory(%q): %v", p.dataDir, err)
-	}
-
-	path := filepath.Join(p.dataDir, pciAddress)
-	err := os.WriteFile(path, []byte(ns), 0600)
+	err := p.store.Set(pciAddress, ns)
 	if err != nil {
-		return fmt.Errorf("failed to write used PCI address lock file in the path(%q): %v", path, err)
+		return fmt.Errorf("Failed to save PCI address %v with ns %v", pciAddress, ns)
 	}
+	return nil
+}
 
-	return err
+func (p *PCIAllocator) LoadAllocatedPCI(pciAddress string) (string, error) {
+	return p.store.Get(pciAddress)
 }
 
-// DeleteAllocatedPCI Remove the allocated PCI file
-// return error if the file doesn't exist
 func (p *PCIAllocator) DeleteAllocatedPCI(pciAddress string) error {
-	path := filepath.Join(p.dataDir, pciAddress)
-	if err := os.Remove(path); err != nil {
-		return fmt.Errorf("error removing PCI address lock file %s: %v", path, err)
-	}
-	return nil
+	return p.store.Delete(pciAddress)
 }
 
-// IsAllocated checks if the PCI address file exist
-// if it exists we also check the network namespace still exist if not we delete the allocation
-// The function will return an error if the pci is still allocated to a running pod
-func (p *PCIAllocator) IsAllocated(pciAddress string) (bool, error) {
-	path := filepath.Join(p.dataDir, pciAddress)
-	_, err := os.Stat(path)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return false, nil
-		}
-
-		return false, fmt.Errorf("failed to check for pci address file for %s: %v", path, err)
-	}
+func namespaceExists(namespace string) bool {
+	networkNamespace, err := ns.GetNS(namespace)
+	ret := err != nil
```

Comment:
Shouldn't we check whether err != nil here?

Otherwise GetNS could do something unexpected, and networkNamespace.Close() might cause a  nil pointer dereference.

Maybe:

```
func namespaceExists(namespace string) bool {
	networkNamespace, err := ns.GetNS(namespace)
	if err != nil {
		return false
	}
	defer networkNamespace.Close()
	return true
}
```

Reply:
good catch

Reply:
fixed.